utils/common/src/main/java/org/apache/brooklyn/util/net/ReachableSocketFinder.java

```diff
@@ -69,86 +65,113 @@ public ReachableSocketFinder(Predicate<? super HostAndPort> socketTester, Listen
+                @Override
+                public Optional<HostAndPort> call() {
+                    // Whether the socket was reachable (vs. the result of call, which is whether the repeater is done).
+                    final AtomicBoolean theResultWeCareAbout = new AtomicBoolean();
+                    Repeater.create("socket-reachable")
+                            .limitTimeTo(timeout)
+                            .backoffTo(Duration.FIVE_SECONDS)
+                            .until(new Callable<Boolean>() {
+                                @Override
+                                public Boolean call() throws TimeoutException {
+                                    boolean reachable = socketTester.apply(socket);
+                                    if (reachable) {
+                                        theResultWeCareAbout.set(true);
+                                        return true;
+                                    } else {
+                                        // Run another check if nobody else has completed yet or another task has
+                                        // completed but this one is still in its grace period.
+                                        Stopwatch timerSinceFirst = sinceFirstCompleted.get();
+                                        return timerSinceFirst != null && Duration.FIVE_SECONDS.subtract(Duration.of(timerSinceFirst)).isNegative();
+                                    }
+                                }
+                            })
+                            .run();
+                    if (theResultWeCareAbout.get()) {
+                        sinceFirstCompleted.compareAndSet(null, Stopwatch.createStarted());
+                    }
+                    return theResultWeCareAbout.get() ? Optional.of(socket) : Optional.<HostAndPort>absent();
+                }
+            }));
+        }
+
+        return new Iterable<Optional<HostAndPort>>() {
+            @Override
+            public Iterator<Optional<HostAndPort>> iterator() {
+                return new AbstractIterator<Optional<HostAndPort>>() {
+                    int count = 0;
+
                     @Override
-                    public void run() {
-                        try {
-                            if (socketTester.apply(socket)) {
-                                reachableSocket.compareAndSet(null, socket);
-                                latch.countDown();
+                    protected Optional<HostAndPort> computeNext() {
+                        if (count < futures.size()) {
+                            final Future<Optional<HostAndPort>> future = futures.get(count++);
+                            try {
+                                return future.get(timeout.toUnit(TimeUnit.MILLISECONDS), TimeUnit.MILLISECONDS);
+                            } catch (Exception e) {
+                                Exceptions.propagateIfInterrupt(e);
```

Comment:
`e` is an `Exception` rather than a `Throwable` so I don't think this is a problem.